Simplifying initialization logic

rjatwal opened the pull request:
Removed `storage_init` hook and allow the extension's `init` hook to initialize the default database. This is necessary to ensure that new connections land in the right database. 

I worry about hidden dependencies on a database being fully initialized before the extension's init is called. I have noticed that catalog lookups do query the default database when they are not fully specified, this can cause issues in existing `init` methods. Not sure whether it makes more sense to reevaluate each hidden dependency to move un-need reliance on the default database, or to add specially opt-in during initialization (to ensure only advanced users use this feature).

rjatwal committed:
Use the duckdb storage but extend the duckdb catalog

rjatwal committed:
Replacing storage_init in creation

rjatwal committed:
Removing fragment of old code

rjatwal committed:
Revert unrelated change

rjatwal committed:
Format fix

rjatwal committed:
Removing obsolete test

Mytherin reviewed (commented):
Thanks for the PR! Great to get rid of this again. Indeed perhaps we do need to have an opt-in - most extensions won't define custom databases/catalogs and should not have to deal with this (e.g. what happens if you do `duckdb httpfs:file.db`). Perhaps we can have a function such as `[ext]_is_storage` or similar.

I will have a go to see if this works with the SQLite extension or if more changes are required there.

rjatwal commented on src/main/database.cpp:
```diff
@@ -199,52 +199,31 @@ void DatabaseInstance::Initialize(const char *database_path, DBConfig *user_conf
-		for (auto &unrecognized_option : config.options.unrecognized_options) {
-			auto entry = config.extension_parameters.find(unrecognized_option.first);
-			if (entry == config.extension_parameters.end()) {
-				throw InvalidInputException("Unrecognized configuration property \"%s\"", unrecognized_option.first);
```
👍

rjatwal replied:
Updated

Mytherin commented:
I have tried this PR and it works for the SQLite extension with the following diff so that aliases are correctly resolved (e.g. `sqlite` gets turned into `sqlite_scanner`, or `md` gets turned into `motherduck`):

```diff
--- a/src/main/database.cpp
+++ b/src/main/database.cpp
@@ -150,7 +150,8 @@ duckdb::unique_ptr<AttachedDatabase> DatabaseInstance::CreateAttachedDatabase(At
        duckdb::unique_ptr<AttachedDatabase> attached_database;
        if (!type.empty()) {
                // find the storage extensionon database
-               auto entry = config.storage_extensions.find(type);
+               auto extension_name = ExtensionHelper::ApplyExtensionAlias(type);
+               auto entry = config.storage_extensions.find(extension_name);
                if (entry == config.storage_extensions.end()) {
                        throw BinderException("Unrecognized storage type \"%s\"", type);
                }

```

rjatwal committed:
PR Feedback

rjatwal committed:
Add alias to attach database lookup

rjatwal commented:
> I have tried this PR and it works for the SQLite extension with the following diff so that aliases are correctly resolved (e.g. `sqlite` gets turned into `sqlite_scanner`, or `md` gets turned into `motherduck`):
> 
> ```diff
> --- a/src/main/database.cpp
> +++ b/src/main/database.cpp
> @@ -150,7 +150,8 @@ duckdb::unique_ptr<AttachedDatabase> DatabaseInstance::CreateAttachedDatabase(At
>         duckdb::unique_ptr<AttachedDatabase> attached_database;
>         if (!type.empty()) {
>                 // find the storage extensionon database
> -               auto entry = config.storage_extensions.find(type);
> +               auto extension_name = ExtensionHelper::ApplyExtensionAlias(type);
> +               auto entry = config.storage_extensions.find(extension_name);
>                 if (entry == config.storage_extensions.end()) {
>                         throw BinderException("Unrecognized storage type \"%s\"", type);
>                 }
> ```
Added

rjatwal committed:
Restoring unrecognized option test

bleskes commented:
>  most extensions won't define custom databases/catalogs and should not have to deal with this (e.g. what happens if you do duckdb httpfs:file.db). Perhaps we can have a function such as [ext]_is_storage or similar.

I think I'm missing something, which is why I'm asking - if an extension doesn't install a storage manager, wouldn't this result in exactly the same thing?  (i.e., an error) - which is what we want? 


what do we expect to happen when people do `duckdb httpfs:file.db`? I think we want an error, but I'm not sure that's what happens in the PR? (we do `CreateDatabase("")`, which if I'm not wrong reading the diff, will just ignore the `httpfs` part - I think we either want a file to created called `httpfs:filedb` or throw an error)

Mytherin commented:
> > most extensions won't define custom databases/catalogs and should not have to deal with this (e.g. what happens if you do duckdb httpfs:file.db). Perhaps we can have a function such as [ext]_is_storage or similar.
> 
> I think I'm missing something, which is why I'm asking - if an extension doesn't install a storage manager, wouldn't this result in exactly the same thing? (i.e., an error) - which is what we want?
> 
> what do we expect to happen when people do `duckdb httpfs:file.db`? I think we want an error, but I'm not sure that's what happens in the PR? (we do `CreateDatabase("")`, which if I'm not wrong reading the diff, will just ignore the `httpfs` part - I think we either want a file to created called `httpfs:filedb` or throw an error)

Yes, you are absolutely correct - after trying the code I realized this as well. Any extension that does not define a storage extension will fail in the `storage_extensions` lookup. Loading a file with `httpfs:file.db` will load the `httpfs` module, then try to attach the `httpfs` storage extension, which does not exist, and then throw an error because the storage extension could not be found. That indeed seems like expected behavior to me so the `[ext]_is_storage` callback is not required

Apologies @rjatwal can we remove the `ExtensionHelper::IsStorageExtension` again?

rjatwal committed:
Removing is_storage guard. Adding autoinstalll onto load extension path

rjatwal committed:
fix parity

rjatwal commented:
Updated PR. I moved the autoinstall into the load extension function. The effect is that the `load extension` will install allow-listed extensions if not already installed. I think this behavior is reasonable and preferred over adding a function to break the parity between `load extension` and initialization code.

Mytherin commented:
Thanks!
